src/btree/bt_delete.c

```diff
@@ -293,19 +293,43 @@ __wt_delete_page_rollback(WT_SESSION_IMPL *session, WT_REF *ref)
  *     IDs in the delete info. Since we're called at the end of recovery there's no need to lock the
  *     ref or worry about races.
  */
-static void
+static int
 __delete_redo_window_cleanup_internal(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_REF *child;
+    uint64_t sleep_usecs, yield_count;
+    bool busy;
 
     WT_ASSERT(session, F_ISSET(ref, WT_REF_FLAG_INTERNAL));
     if (ref->page != NULL) {
         WT_INTL_FOREACH_BEGIN (session, ref->page, child) {
-            if (child->state == WT_REF_DELETED && child->page_del != NULL)
+            /*
+             * We used to do the cleanup when the state of the child was WT_REF_DELETED, but it can
+             * also be loaded in memory if the page has been read before, thus we always clean the
+             * page_del structure if it has data.
+             */
+            if (child->page_del != NULL) {
                 __cell_redo_page_del_cleanup(session, ref->page->dsk, child->page_del);
+                /* Move back the page to WT_REF_DELETED by marking it for eviction. */
+                if (child->state == WT_REF_MEM) {
+                    for (sleep_usecs = yield_count = 0;;) {
+                        WT_RET(__wt_hazard_set(session, child, &busy));
+                        if (!busy) {
+                            WT_RET_BUSY_OK(__wt_page_release_evict(session, child, 0));
+                            break;
```

Comment:
If we evict the page, the page will be having a new write generation number. Will we clear the txnid on this page after recovery or this page must be clean so nothing is changed?

Reply:
If latter is the case, can we add an assert?

Reply:
Does this work with `fast_truncate_2022`?

Reply:
Yes, it works for the `fast_truncate_2022`. We write the `page_del` information to the disk due to the flag is enabled. This change is to reset the `page_del` information after the page is read into memory as part of the recovery.

Reply:
I think this code not only apply to recovery but also the normal code path. Can we try to evict only in recovery? What if rts makes the page dirty, will that cause any problem?

Reply:
`__delete_redo_window_cleanup_internal` is only called during recovery:
`__wt_txn_recover` -> `__wt_dhandle_update_write_gens` -> `__wt_delete_redo_window_cleanup` -> `__delete_redo_window_cleanup_internal`.

For the second question, I am not sure, @kommiharibabu, do you know?

Reply:
I still think we need to add that check. Otherwise, there is an assumption for this function and we may used it wrongly in the future.

Reply:
RTS instantiates the deleted page only when it finds that it is unstable. And yes it marks the page as dirty by removing all the tombstones present on the page. Evicting that page doesn't cause any problems. Later reading this page will not have any tombstones.

Reply:
I added a check in 01cf5d0b21af378d275db099a41ae5ddc16ebcee at the start of the function.